lib/graphs/digraph.nit

```diff
@@ -0,0 +1,606 @@
+# Implementations for representing directed graphs, also called digraphs.
+#
+# Currently, there are two classes
+module digraph
+
+# An arc of a digraph
+class Arc[V]
+	# The source of the arc
+	#
+	# ~~~
+	# var arc = new Arc[Int](0, 1)
+	# assert arc.source == 0
+	# ~~~
+	var source: V
+
+	# The target of the arc
+	#
+	# ~~~
+	# var arc = new Arc[Int](0, 1)
+	# assert arc.target == 1
+	# ~~~
+	var target: V
+
+	redef fun to_s: String
+	do
+		return "({source.to_s}, {target.to_s})"
+	end
+end
+
+# Interface for digraphs
+abstract class AbstractDigraph[V: Object]
+
+	## ---------- ##
+	## Properties ##
+	## ---------- ##
+
+	# The number of vertices in this graph.
+	#
+	# ~~~
+	# var g = new HashMapDigraph[Int]
+	# g.add_vertex(0)
+	# g.add_vertex(1)
+	# assert g.num_vertices == 2
+	# g.add_vertex(0)
+	# assert g.num_vertices == 2
+	# ~~~
+	var num_vertices = 0
+
+	# The number of arcs in this graph.
+	# ~~~
+	# var g = new HashMapDigraph[Int]
+	# g.add_arc(0, 1)
+	# assert g.num_arcs == 1
+	# g.add_arc(0, 1)
+	# assert g.num_arcs == 1
+	# g.add_arc(2, 3)
+	# assert g.num_arcs == 2
+	# ~~~
+	var num_arcs = 0
```

Comment:
Maybe `num_vertices` and `num_arcs` should be abstract methods, and the subclasses should have to provide specific implementation (or just `redef var num_arc = 0` if they just want an attribute).
But I think it is better to let the choice of the implementation to subclasses.

For instance, in the 2 subclasses `num_vertices` could be implemented respectively by

~~~nit
redef fun num_vertices do return pred_map.length
redef fun num_vertices do return vertex_to_index.length
~~~